pkg/cmd/server/start/start_master.go

```diff
@@ -265,11 +265,24 @@ func (o MasterOptions) RunMaster() error {
 		return kerrors.NewInvalid("MasterConfig", o.ConfigFile, validationResults.Errors)
 	}
 
-	if err := StartMaster(masterConfig); err != nil {
-		return err
+	if !o.MasterArgs.StartControllers {
+		masterConfig.Controllers = configapi.ControllersDisabled
 	}
 
-	return nil
+	if !o.MasterArgs.StartAPI {
+		openshiftConfig, err := origin.BuildMasterConfig(*masterConfig)
+		if err != nil {
+			return err
+		}
+
+		_, kubeConfig, err := buildKubernetesMasterConfig(openshiftConfig)
+		if err != nil {
+			return err
+		}
+		return StartControllers(openshiftConfig, kubeConfig)
```

Comment:
I don't like the way this logic is built.  Why not simply build the `masterConfig` with correct values and have `StartMaster(*origin.MasterConfig)` start the correct pieces as it goes?  That seems more maintainable if we're going to use this as the common entry point.